core/src/main/java/software/amazon/awssdk/core/pagination/async/PaginatedItemsPublisher.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.pagination.async;
+
+import java.util.Iterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+import org.reactivestreams.Subscriber;
+
+/**
+ * A publisher to request for a stream of paginated items. The class can be used to request data for paginated items
+ * across multiple pages.
+ *
+ * @param <ResponseT> The type of a single response page
+ * @param <ItemT> The type of paginated member in a response page
+ */
+public class PaginatedItemsPublisher<ResponseT, ItemT> implements SdkPublisher<ItemT> {
```

Comment:
There's a lot of duplicated code from the Response publisher. Is there a way to consolidate?

This code from Rx might be relevant for making this more abstract

https://github.com/ReactiveX/RxJava/blob/2.x/src/main/java/io/reactivex/internal/operators/flowable/FlowableGenerate.java

Reply:
Didn't use the above link. But abstracted out the common code.